### PR TITLE
issue #41: improve the speed when browse Tests with JDT Search Engine

### PR DESCRIPTION
**Pull request by missedone**
fix issue #41 by using JDT Search Engine to search "Test" annotation in project source files.

### CODE REVIEW

File: testng-eclipse-plugin/src/main/org/testng/eclipse/launch/GroupSelector.java [hunk 1]
```diff
@@ -55,16 +58,16 @@
...
 
-      Object[] projects = new Object[1 + dependencies.length];
+      IJavaProject[] projects = new IJavaProject[1 + dependencies.length];
       projects[0] = selectedProject;
       System.arraycopy(dependencies, 0, projects, 1, dependencies.length);
-      Object[] types = TestSearchEngine.findTests(getCallback().getLaunchConfigurationDialog(),
-          projects, Filters.SINGLE_TEST);
+      Set<IType> types = new HashSet<>();
+      for (IJavaProject project : projects) {
+        types.addAll(Arrays.asList(TestSearchEngine.findTestNGTests(getCallback().getLaunchConfigurationDialog(),
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/launch/GroupSelector.java]: use the new method `TestSearchEngine.findTestNGTests` which facilitate with JDT search engine

File: testng-eclipse-plugin/src/main/org/testng/eclipse/launch/SuiteSelector2.java [hunk 2]
```diff
@@ -6,7 +6,7 @@
 import java.util.List;
 import java.util.Map;
 
-import org.eclipse.core.internal.resources.File;
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/launch/SuiteSelector2.java]: changes in this file is to removing using internal API

File: testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestFinder.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+      IRegion region= CoreTestSearchEngine.getRegion(element);
+      ITypeHierarchy hierarchy= JavaCore.newTypeHierarchy(region, null, new SubProgressMonitor(pm, 1));
+      IType[] allClasses= hierarchy.getAllClasses();
+
+      // search for all types with references to RunWith and Test and all subclasses
+      HashSet<IType> candidates= new HashSet<>(allClasses.length);
+      SearchRequestor requestor= new AnnotationSearchRequestor(hierarchy, candidates);
+
+      IJavaSearchScope scope= SearchEngine.createJavaSearchScope(allClasses, IJavaSearchScope.SOURCES);
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestFinder.java]: the benefit of JDT search engine is we can easily specify the Search scope as SOURCES and we can benefit from the indexing of JDT search

**Comment by missedone** [comment]
hi @cbeust 

here is some numbers of the duration of searching 38 test classes out of 154 java source files (under same java project), each line indicates one click on "Browse" button (10 times in total):
1. original
```
>>>>>>>>> found 40 tests spent time: 8589
>>>>>>>>> found 40 tests spent time: 4601
>>>>>>>>> found 40 tests spent time: 3847
>>>>>>>>> found 40 tests spent time: 3664
>>>>>>>>> found 40 tests spent time: 3592
>>>>>>>>> found 40 tests spent time: 3528
>>>>>>>>> found 40 tests spent time: 3547
>>>>>>>>> found 40 tests spent time: 3626
>>>>>>>>> found 40 tests spent time: 3571
>>>>>>>>> found 40 tests spent time: 3881
```
2. with PR:
```
>>>>>>>>> found 38 tests spent time: 6126
>>>>>>>>> found 38 tests spent time: 1379
>>>>>>>>> found 38 tests spent time: 1143
>>>>>>>>> found 38 tests spent time: 992
>>>>>>>>> found 38 tests spent time: 911
>>>>>>>>> found 38 tests spent time: 843
>>>>>>>>> found 38 tests spent time: 824
>>>>>>>>> found 38 tests spent time: 805
>>>>>>>>> found 38 tests spent time: 812
>>>>>>>>> found 38 tests spent time: 777
>>>>>>>>> found 38 tests spent time: 771
```

as you can see, the PR take advantage of Eclipse JDT search engine, the search speed is benefit with the indexing, so it take less and less time to list all test classes.
the numbers of original code trend to be stable around 3 secs

File: testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestFinder.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+import org.eclipse.jdt.core.search.IJavaSearchConstants;
+import org.eclipse.jdt.core.search.IJavaSearchScope;
+import org.eclipse.jdt.core.search.SearchEngine;
+import org.eclipse.jdt.core.search.SearchParticipant;
+import org.eclipse.jdt.core.search.SearchPattern;
+import org.eclipse.jdt.core.search.SearchRequestor;
+import org.eclipse.jdt.internal.junit.util.CoreTestSearchEngine;
+import org.testng.eclipse.util.ResourceUtil;
+
+public class TestFinder {
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestFinder.java]: this TestFinder is copied from JUnit Eclipse plugin, and slightly modified to adapt our needs

File: testng-eclipse-plugin/src/main/org/testng/eclipse/util/TestSearchEngine.java [hunk 5]
```diff
@@ -7,9 +7,6 @@
 import java.util.Map;
 import java.util.Set;
 
-import com.google.common.collect.Sets;
-
-import org.eclipse.core.internal.resources.File;
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/util/TestSearchEngine.java]: prefer public API

**Comment by missedone** [comment]
one thing you might curious is why the original code found 40 test classed but the code with PR only found 38, this is because there're 2 inner classes in one of the test classes, the inner classes are not test class (is a) at all, but only used by the test class (has a).

**Comment by cbeust** [comment]
That's awesome, @missedone. Both the PR and how you document it. Thanks!

**Pull request merged**